### PR TITLE
migrated from Fsyom to QStateMachine

### PR DESCRIPTION
**Pull request by t0mpr1c3**
* migrated from Fsyom to QStateMachine
* deleted empty plugins
* removed race condition by adding FINISHING State
* fixed menu behaviour
* moved graphics scene methods to separate module

### CODE REVIEW

**Comment by t0mpr1c3** [comment]
Referencing https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/338

**Comment by t0mpr1c3** [comment]
Referencing https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/345#issuecomment-646997698 (Configure button)

**Comment by t0mpr1c3** [comment]
simpleaudio requires asoundlib.h (`apt install libasound2-dev`)

**Comment by t0mpr1c3** [comment]
Chris, this is ready to go now that I have fixed the audio library dependency.

File: src/main/python/ayab/ayab_fsm.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+class FSM (object):
+    """Heirarchical Finite State Machine.
+
+    This replaces `fsyom` and various imperatively-programmed structures
+    with a declaratively-programmed, event-driven framework tailored to the UI.
+    @author Tom Price
+    @date   June 2020
+    """
+    def __init__(self, parent):
+        """Define Heirarchical states for Finite State Machine"""
```
Comment (chris007de) [inline src/main/python/ayab/ayab_fsm.py:36]: typo

File: src/main/python/ayab/ayab.py [hunk 2]
```diff
@@ -549,14 +358,19 @@ def getSerialPorts(self):
...
-        # if event == "start":
-        #     playsound(self.app_context.get_resource("assets/start.wav"))
-        # if event == "nextline":
-        #     playsound(self.app_context.get_resource("assets/nextline.wav"))
-        # if event == "finished":
-        #     playsound(self.app_context.get_resource("assets/finish.wav"))
+    def playsound(self, sound):
+        if str2bool(self.prefs.settings.value("quiet_mode")):
+            return 
+        dirname = path.join("..", "..", "resources", "base", "assets", path.dirname(__file__))
```
Comment (chris007de) [inline src/main/python/ayab/ayab.py:364]: When running the application, the .wav files are not found. (`WARNING  File start.wav not found.`)
Please use self.app_context.get_resource() (see before your commit) to get an OS independent location of the files.

File: src/main/python/ayab/ayab.py [hunk 3]
```diff
@@ -549,14 +358,19 @@ def getSerialPorts(self):
...
-        # if event == "start":
-        #     playsound(self.app_context.get_resource("assets/start.wav"))
-        # if event == "nextline":
-        #     playsound(self.app_context.get_resource("assets/nextline.wav"))
-        # if event == "finished":
-        #     playsound(self.app_context.get_resource("assets/finish.wav"))
+    def playsound(self, sound):
+        if str2bool(self.prefs.settings.value("quiet_mode")):
+            return 
+        dirname = self.app_context.get_resource("base/assets")
```
Comment (chris007de) [inline src/main/python/ayab/ayab.py:364]: Needs to be: `dirname = self.app_context.get_resource("assets")`
Could have told you that in my first comment already

**Comment by chris007de** [comment]
* For me, the progress window always starts in the 2nd or 3rd row. Console log shows correct rows.
* Additionally, cancelling knitting and switching to another pattern shows the previous pattern in the progress window.

**Comment by t0mpr1c3** [comment]
> * For me, the progress window always starts in the 2nd or 3rd row. Console log shows correct rows.
> * Additionally, cancelling knitting and switching to another pattern shows the previous pattern in the progress window.

1. I set a 2 second delay in simulation mode before knitting starts to try to avoid that, but it would be better to require an acknowledgement signal.
2. The knit progress continues to show after cancel has been pressed, but does go away when Knit is pressed a second time. Do you think it should disappear sooner than that?

**Comment by chris007de** [comment]
> > * For me, the progress window always starts in the 2nd or 3rd row. Console log shows correct rows.
> > * Additionally, cancelling knitting and switching to another pattern shows the previous pattern in the progress window.
> 
> 1. I set a 2 second delay in simulation mode before knitting starts to try to avoid that, but it would be better to require an acknowledgement signal.

Ok, so I understand that it is an issue of the simulation mode only. We'll address this in a separate issue.

> 2. The knit progress continues to show after cancel has been pressed, but does go away when Knit is pressed a second time. Do you think it should disappear sooner than that?

The way the progress data appears and disappears is fine for me. What I wanted to say is that if I knit pattern A, cancel, load pattern B, knit, still the data of pattern A is shown in the progress window. Please try if you can reproduce this on your side.
we can address this in a separate issue as well.

**Comment by t0mpr1c3** [comment]
Thanks, I implemented the `get_resource()` call. I haven't been able to reproduce the second problem yet.

**Comment by t0mpr1c3** [comment]
I fixed bug #1 by synchronizing the threads at the "Please Knit" notification. I still can't reproduce bug #2 at the moment.

**Pull request merged**